### PR TITLE
Start using React context to share editor instance

### PR DESCRIPTION
**Pull request by wincent**
Instead of passing the editor through props everywhere, we can get it out of props and into context. This will make dealing with #1037 easier, I hope, because it will be one less prop we have to worry about propagating through our HOCs; but I think it makes sense to do this anyway because the editor is a "global" value ("global" per editor instance, that is), which is exactly the use case that context exists for.

First commit shows using context in a class component at the top of the hierarchy. Second commit shows using context in a functional component in a leaf at the bottom of the hierarchy. I'll then proceed to add more commits until the whole tree is ported. At each step, my test plan is `npm run dev && npm run test && npm run start` and checking the demo.

### CODE REVIEW

File: src/components/buttons/button-accessibility-image-alt.jsx [hunk 1]
```diff
@@ -27,10 +30,10 @@ class ButtonAccessibilityImageAlt extends React.Component {
 		editor: PropTypes.object.isRequired,
 	};
 
-	constructor(prop, context) {
-		super(prop, context);
+	constructor(prop) {
+		super(prop);
```
Comment (wincent) [inline src/components/buttons/button-accessibility-image-alt.jsx]: Constructors don't receive `context` (that was part of the legacy context API), and in any case, I don't think we were ever using it here.

**Comment by wincent** [comment]
I think this is ready for review now @julien.

**Comment by wincent** [comment]
Rebased onto current `2.x-develop`.

**Comment by wincent** [comment]
Tested this pretty thoroughly in demo and seems to work in portal too:

new_web_content_-_liferay

File: src/adapter/editor-context.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	const componentWithContext = props => (
+		<EditorContext.Consumer>
+			{context => <Component {...props} context={context} />}
+		</EditorContext.Consumer>
+	);
+	componentWithContext.key = Component.key;
+	return componentWithContext;
+};
+
+export default EditorContext;
```
Comment (jbalsas) [inline src/adapter/editor-context.js]: Not sure where we stand on this from our previous conversations on other projects... not to worry here, but should we go with named exports everywhere? Thought that was the consensus 🤷‍♀️

File: src/components/base/toolbar-buttons.js [hunk 3]
```diff
@@ -1,16 +1,19 @@
...
  *
  * @class ToolbarButtons
  */
 export default WrappedComponent =>
-	class extends WrappedComponent {
+	class ToolbarButtons extends WrappedComponent {
+		static contextType = EditorContext;
+
 		/**
-		 * Analayses the current selection and returns the buttons or button groups to be rendered.
```
Comment (jbalsas) [inline src/components/base/toolbar-buttons.js]: You can't unseen this 😂

File: src/components/buttons/button-accessibility-image-alt.jsx [hunk 4]
```diff
@@ -14,23 +14,16 @@ import ReactDOM from 'react-dom';
...
-		/**
-		 * The editor instance where the component is being used.
-		 *
-		 * @instance
-		 * @memberof ButtonAccessibilityImageAlt
-		 * @property {Object} editor
-		 */
-		editor: PropTypes.object.isRequired,
-	};
+	static propTypes = {};
```
Comment (jbalsas) [inline src/components/buttons/button-accessibility-image-alt.jsx]: Just out of curiosity, what's the benefit of an empty `proptypes` definition? Looks like we're not validating `propTypes` in many places... should we file an issue to track down what we want to do about it?

Comment (wincent) [reply src/components/buttons/button-accessibility-image-alt.jsx]: I don't think there's any real "benefit". I just wanted to make the minimal change, which was removing the property. You probably won't be surprised to know that I would rather see us using static types and just get rid of runtime prop-type checking.

File: src/adapter/editor-context.js (hunk 2 above)
Comment (wincent) [reply src/adapter/editor-context.js]: Previous discussion was on https://github.com/liferay/clay/pull/1453 — summarizing:

- That PR just uses a default export.
- Some articles were shared arguing in favor of named exports.
- @julien expressed a preference for named exports.
- You noted that, at least in Clay, we pretty strictly follow a one-export-per-module, making default exports simple.
- PR got merged, possibly without enough explicit agreement on what exactly we were agreeing to.

I'm open to sticking to default exports or named exports, and I think the one-export-per-module guideline is a good rule of thumb. Being consistent about it and picking "an" option is probably more valuable than picking the "right" option. It's not too late to change our minds if we think that one was a mistake.

File: src/components/buttons/button-accessibility-image-alt.jsx (hunk 4 above)
Comment (wincent) [reply src/components/buttons/button-accessibility-image-alt.jsx]: Followed up in #1084.

**Pull request merged**